monai/data/dataset.py

```diff
@@ -93,22 +97,64 @@ def __init__(self, data, transform, cache_num=sys.maxsize, cache_rate=1.0):
                 will take the minimum of (cache_num, data_length x cache_rate, data_length).
             cache_rate (float): percentage of cached data in total, default is 1.0 (cache all).
                 will take the minimum of (cache_num, data_length x cache_rate, data_length).
+            num_workers (int): the number of worker processes to use.
+                If 0 a single thread will be used. Default is 0.
         """
         if not isinstance(transform, Compose):
             transform = Compose(transform)
         super().__init__(data, transform)
         self.cache_num = min(cache_num, int(len(self) * cache_rate), len(self))
-        self._cache = list()
+        self._cache = [None] * self.cache_num
         print('Load and cache transformed data...')
-        for i in range(self.cache_num):
-            process_bar(i + 1, self.cache_num)
-            item = data[i]
-            for _transform in transform.transforms:
-                # execute all the deterministic transforms before the first random transform
-                if isinstance(_transform, Randomizable):
-                    break
-                item = apply_transform(_transform, item)
-            self._cache.append(item)
+        if num_workers > 0:
+            self._item_processed = 0
+            if mode == "thread":
+                lock = Lock()
+
+                def _update_bar(arg):
+                    self._item_processed += 1
+                    lock.acquire()
```

Comment:
Here you can use `lock` as a context manager rather than try/finally, so 

```
with lock:
    process_bar(self._item_processed, self.cache_num)
```

Reply:
Thanks, good point. I was also wondering if I should rather call `lock.acquire()` with blocking argument set to `False`, so that if the process bar is being updated in one thread, another thread doesn't have to wait for it and simply skips updating it. The only downside I see is if the thread that adds the last item to the cache "skips" updating the bar as it is already being updated in another thread, the bar might stay "unfilled". What do you think?

Reply:
I don't think it's necessary to optionally lock, updating the process bar should be fast enough to not worry about the odd time waiting would occur. I would suggest the context block syntax as it's much less error-prone.